### DISCUSSION TITLE
BindableCollection<T>.RemoveRange slow

### DISCUSSION THREAD

**Discussion by GF-Huang** [canton7/Stylet#181] [Q&A] [answered]
Only one element to remove, take 200+ms.

![image](https://user-images.githubusercontent.com/4510984/102022355-4d4ff400-3dc1-11eb-8fe0-008001d2c4fe.png)

**Comment by canton7** [comment, marked as answer]
Bear in mind, that number includes the cost of line 35, and also all of the UI actions which happen as a result of firing the `CollectionChanged` event, with a type of `Reset`. In practice, this probably means re-rendering the UI collection.

I'm not really sure what I'm supposed to add here I'm afraid -- you can see the code for `BindableCollection` in this repo, and it's clear there's no `Thread.Sleep(287)`. Therefore the cost is coming from the code which handles the `CollectionChanged` event, obviously.

**Reply by GF-Huang** [reply]
No, it is not includes line 35, it takes only 3ms.

![6q5cHgvYp3](https://user-images.githubusercontent.com/4510984/102022759-e849cd80-3dc3-11eb-8a24-886e0376f14a.gif)

**Reply by canton7** [reply]
Therefore the cost is coming from the UI processing the `CollectionChanged` event, which is what the rest of my last comment talked about. (There's also potentially the cost of moving to the UI thread, if you're not on it already)

**Reply by canton7** [reply]
If your collection is large and the number of items to remove is small, you might find it cheaper to call `.Remove` multiple times, rather than `.RemoveRange`. This will fire potentially more `CollectionChanged` events, but they'll be of type `Remove`, which will allow the UI to remove those items more efficiently.

**Reply by GF-Huang** [reply]
It's less than 10 elements 😄, but thanks your suggestion, I will try `.Remove`.

**Reply by canton7** [reply]
Regardless, unless you can show that the slow-down is in Stylet's code and not in your UI, I think this is your issue to investigate. Good luck!
